File: platforms/ios/Sources/Trace/TraceSession.swift
// TraceSession.swift — streaming session for AI-drawn content. Wraps the Rust
// core's `Session` via UniFFI. Mirrors the API in SPEC §4.5.

import Foundation

public final class TraceSession {
    fileprivate let inner: TraceCore.Session

    public init(viewBox: CGRect) {
        self.inner = TraceCore.Session(viewbox: TraceCore.ViewBox(
            x: Float(viewBox.origin.x),
            y: Float(viewBox.origin.y),
            w: Float(viewBox.size.width),
            h: Float(viewBox.size.height)
        ))
    }

    @discardableResult
    public func append(svgFragment: String, anim: FragmentAnim? = nil) throws -> [String] {
        try inner.appendSvgFragment(svg: svgFragment, anim: anim?.toCore())
    }

    public func remove(id: String) -> Bool { inner.removeElement(id: id) }

    public var version: UInt64 { inner.version() }

    public func snapshot() throws -> Data {
        Data(try inner.snapshot())
    }
}

public struct FragmentAnim {
    public var strokeDrawMs: UInt32?
    public var fillFadeMs: UInt32?
    public var startAfter: StartAfter
    public var easing: Easing

    public init(strokeDrawMs: UInt32? = nil,
                fillFadeMs: UInt32? = nil,
                startAfter: StartAfter = .previousFragment,
                easing: Easing = .handNatural) {
        self.strokeDrawMs = strokeDrawMs
        self.fillFadeMs = fillFadeMs
        self.startAfter = startAfter
        self.easing = easing
    }

    fileprivate func toCore() -> TraceCore.FragmentAnim {
        TraceCore.FragmentAnim(
            strokeDrawMs: strokeDrawMs,
            fillFadeMs: fillFadeMs,
            startAfter: startAfter.toCore(),
            easing: easing.toCore()
        )
    }
}

public enum StartAfter {
    case immediately, previousFragment, atMs(UInt32)
    fileprivate func toCore() -> TraceCore.StartAfter {
        switch self {
        case .immediately: return .immediately
        case .previousFragment: return .previousFragment
        case .atMs(let t): return .atMs(t: t)
        }
    }
}

public enum Easing {
    case linear, easeOut, easeInOut, handNatural
    fileprivate func toCore() -> TraceCore.Easing {
        switch self {
        case .linear:       return .linear
        case .easeOut:      return .easeOut
        case .easeInOut:    return .easeInOut
        case .handNatural:  return .handNatural
        }
    }
}

File: platforms/android/src/main/kotlin/co/trace/TraceSession.kt
package co.trace

import android.graphics.RectF
import co.trace.generated.Easing as CoreEasing
import co.trace.generated.FragmentAnim as CoreFragmentAnim
import co.trace.generated.Session as CoreSession
import co.trace.generated.StartAfter as CoreStartAfter
import co.trace.generated.ViewBox as CoreViewBox

/** Streaming session for AI-drawn content. See SPEC §4.5. */
class TraceSession(viewBox: RectF) {
    internal val inner: CoreSession = CoreSession(
        CoreViewBox(viewBox.left, viewBox.top, viewBox.width(), viewBox.height())
    )

    @Throws(Exception::class)
    fun append(svgFragment: String, anim: FragmentAnim? = null): List<String> =
        inner.appendSvgFragment(svgFragment, anim?.toCore())

    fun remove(id: String): Boolean = inner.removeElement(id)

    val version: ULong get() = inner.version()

    @Throws(Exception::class)
    fun snapshot(): ByteArray = inner.snapshot().toByteArray()
}

data class FragmentAnim(
    val strokeDrawMs: UInt? = null,
    val fillFadeMs: UInt? = null,
    val startAfter: StartAfter = StartAfter.PreviousFragment,
    val easing: Easing = Easing.HandNatural,
) {
    internal fun toCore() = CoreFragmentAnim(
        strokeDrawMs = strokeDrawMs,
        fillFadeMs = fillFadeMs,
        startAfter = startAfter.toCore(),
        easing = easing.toCore(),
    )
}

sealed class StartAfter {
    object Immediately : StartAfter()
    object PreviousFragment : StartAfter()
    data class AtMs(val t: UInt) : StartAfter()

    internal fun toCore(): CoreStartAfter = when (this) {
        Immediately       -> CoreStartAfter.Immediately
        PreviousFragment  -> CoreStartAfter.PreviousFragment
        is AtMs           -> CoreStartAfter.AtMs(this.t)
    }
}

enum class Easing {
    Linear, EaseOut, EaseInOut, HandNatural;

    internal fun toCore(): CoreEasing = when (this) {
        Linear      -> CoreEasing.LINEAR
        EaseOut     -> CoreEasing.EASE_OUT
        EaseInOut   -> CoreEasing.EASE_IN_OUT
        HandNatural -> CoreEasing.HAND_NATURAL
    }
}
